monai/transforms/utils.py

```diff
@@ -392,7 +463,53 @@ def _correct_centers(
         center = np.unravel_index(indices_to_use[random_int], label_spatial_shape)
         # shift center to range of valid centers
         center_ori = list(center)
-        centers.append(_correct_centers(center_ori, valid_start, valid_end))
+        centers.append(correct_crop_centers(center_ori, spatial_size, label_spatial_shape))
+
+    return centers
+
+
+def generate_label_classes_crop_centers(
+    spatial_size: Union[Sequence[int], int],
+    num_samples: int,
+    ratios: List[Union[float, int]],
+    label_spatial_shape: Sequence[int],
+    indices: List[np.ndarray],
+    rand_state: Optional[np.random.RandomState] = None,
+) -> List[List[np.ndarray]]:
+    """
+    Generate valid sample locations based on the specified ratios of label classes.
+    Valid: samples sitting entirely within image, expected input shape: [C, H, W, D] or [C, H, W]
+
+    Args:
+        spatial_size: spatial size of the ROIs to be sampled.
+        num_samples: total sample centers to be generated.
+        ratios: ratios of every class in the label to generate crop centers, including background class.
+        label_spatial_shape: spatial shape of the original label data to unravel selected centers.
+        indices: sequence of pre-computed foreground indices of every class in 1 dimension.
+        rand_state: numpy randomState object to align with other modules.
+
+    """
+    if rand_state is None:
+        rand_state = np.random.random.__self__  # type: ignore
+
+    if len(ratios) != len(indices):
+        raise ValueError("random crop radios must match the number of indices of classes.")
+    # ensure indices are numpy array
+    indices = [np.asarray(i) for i in indices]
+    for i, array in enumerate(indices):
+        if len(array) == 0:
+            warnings.warn(f"no available indices of class {i} to crop, set the crop ratio of this class to zero.")
+            ratios[i] = 0
+
+    centers = []
+    for _ in range(num_samples):
+        # randomly select the indices of a class based on the ratios
+        indices_to_use = indices[rand_state.choice(len(ratios), p=np.asarray(ratios) / np.sum(ratios))]
```

Comment:
the `ratios` needs more input validation.
also I think `rand_state.choice(len(ratios), size= num_samples, p=np.asarray(ratios) / np.sum(ratios))` before the for loop will be slightly more efficient

Reply:
added non-negative check for ratios.

I think every sample should:
(1) randomly choose which class to select center.
(2) randomly select a center to crop.

So we can't move the `rand_state.choice(len(ratios), size= num_samples, p=np.asarray(ratios) / np.sum(ratios))` before the for loop.

Thanks.

Reply:
I think for (1) generating one sample within the loop is equivalent to generating 'num_samples' before the loop and indexing the samples

Reply:
Oh, yes, you are right.
I updated in the latest commit.

Thanks.